src/cmd/tools/verify_data_files/main/main.go

```diff
@@ -0,0 +1,473 @@
+		zap.Strings("namespaces", namespaces))
+	var fileSetFiles []fs.FileSetFile
+	for _, namespace := range namespaces {
+		namespacePath := path.Join(dataDirPath, namespace)
+		shards, err := dirFiles(namespacePath)
+		if err != nil {
+			log.Fatal("could not read shards for namespace",
+				zap.String("namespacePath", namespacePath),
+				zap.Error(err))
+		}
+
+		log.Debug("discovered shards",
+			zap.String("namespace", namespace),
+			zap.String("namespacePath", namespacePath),
+			zap.Strings("shards", shards))
+		for _, shard := range shards {
+			shardPath := path.Join(namespacePath, shard)
+			shardID, err := strconv.Atoi(shard)
+			if err != nil {
+				log.Fatal("could not parse shard dir as int",
+					zap.String("shardPath", shardPath), zap.Error(err))
+			}
+
+			shardFileSets, err := fs.DataFiles(filePathPrefix,
+				ident.StringID(namespace), uint32(shardID))
+			if err != nil {
+				log.Fatal("could not list shard dir file setes",
+					zap.String("shardPath", shardPath), zap.Error(err))
+			}
+
+			log.Debug("discovered shard file sets",
+				zap.String("namespace", namespace),
+				zap.String("namespacePath", namespacePath),
+				zap.Int("shardID", shardID),
+				zap.Any("fileSets", shardFileSets))
+			fileSetFiles = append(fileSetFiles, shardFileSets...)
+		}
+	}
+
+	// Sort by time in reverse (usually want to fix latest files first and
+	// can stop once done with fail-fast).
+	log.Info("sorting file sets", zap.Int("numFileSets", len(fileSetFiles)))
+	sort.Slice(fileSetFiles, func(i, j int) bool {
+		return fileSetFiles[i].ID.BlockStart.After(fileSetFiles[j].ID.BlockStart)
+	})
+
+	log.Info("verifying file sets", zap.Int("numFileSets", len(fileSetFiles)))
+	for _, fileSet := range fileSetFiles {
+		log.Info("verifying file set file", zap.Any("fileSet", fileSet))
+		if err := verifyFileSet(verifyFileSetOptions{
+			filePathPrefix: filePathPrefix,
+			bytesPool:      bytesPool,
+			fileSet:        fileSet,
+			fixDir:         opts.fixDir,
+			fixInvalidIDs:  opts.fixInvalidIDs,
+			fixInvalidTags: opts.fixInvalidTags,
+		}, log); err != nil {
+			log.Error("file set file failed verification",
+				zap.Error(err),
+				zap.Any("fileSet", fileSet))
+
+			if opts.failFast {
+				log.Fatal("aborting due to fail fast set")
+			}
+		}
+	}
+}
+
+func dirFiles(dirPath string) ([]string, error) {
+	dir, err := os.Open(dirPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not open dir: %v", err)
+	}
+
+	defer dir.Close()
+
+	stat, err := dir.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("could not stat dir: %v", err)
+	}
+	if !stat.IsDir() {
+		return nil, fmt.Errorf("path is not a directory: %s", dirPath)
+	}
+
+	entries, err := dir.Readdirnames(-1)
+	if err != nil {
+		return nil, fmt.Errorf("could not read dir names: %v", err)
+	}
+
+	results := make([]string, 0, len(entries))
+	for _, p := range entries {
+		if p == "." || p == ".." || p == "./.." || p == "./" || p == "../" || p == "./../" {
```

Comment:
can you do `if p[0] == '.' { continue } `? Or is there a chance there'll be valid files starting with `.`?

Reply:
Correct they could be "./foo"